source/agora/consensus/Reward.d

```diff
@@ -0,0 +1,413 @@
+
+    public this (in uint payout_period, in Duration block_interval)
+    {
+        this.log = Logger(__MODULE__);
+
+        this.payout_period = payout_period;
+        this.block_interval_sec = block_interval.total!"seconds";
+
+        // Calculate payouts per year
+        this.payouts_per_year = yearOfSecs / (this.payout_period * this.block_interval_sec);
+
+        // Calculate amount that would be paid to Commons Budget for full payout period
+        // if the allocation is still sufficient
+        this.commonsPerPayout = commonsCoinsPerBlock; commonsPerPayout.mul(this.payout_period);
+    }
+
+    /***************************************************************************
+
+        Returns the Block Rewards for the given height.
+
+        Params:
+            height = last block of the current payout period
+                (the CoinBase transaction will be in block at this height plus the payout period)
+            percent_signed = (#actual signatures / #expected signatures) * 100
+                on all the blocks for payout period
+
+        Returns:
+            `BlockRewards` which indicates the `Amount` to pay the `Validators`
+            and the `Commons Budget`
+
+    ***************************************************************************/
+
+    public BlockRewards calculateBlockRewards (in Height height, in ubyte percent_signed) nothrow @safe
+    {
+        assert(height % this.payout_period == 0, "This is not a payout block height");
+        assert (height > 0, "We can not payout in Genesis block");
+
+        auto validator_allocated = allocatedValRewards(height);
+        auto validator = reducedValRewards(validator_allocated, percent_signed);
+        auto leftover = validator_allocated; leftover.sub(validator);
+        auto commons = commonsBudgetRewards(height); commons.add(leftover);
+
+        return BlockRewards(validator, commons);
+    }
+
+    /***************************************************************************
+
+        Returns the Commons Budget reward as defined in the Whitepaper.
+
+        Params:
+            height = block height of coinbase payout block
+                (checked to be valid in `Reward` constructor)
+
+        Returns:
+            Reward amount
+
+    ***************************************************************************/
+
+    private Amount commonsBudgetRewards (in Height height) nothrow @safe
+    {
+        // Calculate payout number this one will be
+        auto payouts = height / this.payout_period;
+
+        // Sum up all the rewards so far
+        Amount rewards = this.commonsPerPayout; rewards.mul(payouts);
+
+        // If the allocation is all paid out
+        try
+        {
+            ensure(rewards < this.totalCommonsBudgetRewards,
+                "No more rewards for Commons Budget at height {}", height);
+        }
+        catch (Exception e)
+        {
+            log.trace(e.msg);
+            return 0.coins;
+        }
```

Comment:
What ?

Reply:
The allocation of Commons Budget rewards are limited and will be paid out till some time in year 6. I'll add a comment in the code also.

Reply:
Or did you not like the use of the Exception?

Reply:
I don't understand why we would have `ensure(cond)` inside a `try` / `catch` instead of `if (!cond) catchBody` ?

Reply:
Me neither.